cpp/src/arrow/json/parser_benchmark.cc

```diff
@@ -30,24 +30,50 @@
 namespace arrow {
 namespace json {
 
-std::shared_ptr<Schema> TestSchema() {
-  return schema({field("int", int32()), field("str", utf8())});
-}
-
-constexpr int seed = 0x432432;
-
-std::string TestJsonData(int num_rows, bool pretty = false) {
-  std::default_random_engine engine(seed);
-  std::string json;
-  for (int i = 0; i < num_rows; ++i) {
-    StringBuffer sb;
-    Writer writer(sb);
-    ABORT_NOT_OK(Generate(TestSchema(), engine, &writer));
-    json += pretty ? PrettyPrint(sb.GetString()) : sb.GetString();
-    json += "\n";
+class JSONGenerator {
+ public:
+  constexpr static int kSeed = 0x432432;
+
+  constexpr explicit JSONGenerator(bool pretty = false) : pretty_(pretty) {}
+
+  template <typename T>
+  std::string operator()(const T& input, int32_t num_rows) const {
+    std::default_random_engine engine(kSeed);
+    std::string json;
+    for (int i = 0; i < num_rows; ++i) {
+      StringBuffer sb;
+      Writer writer(sb);
+      ABORT_NOT_OK(Generate(input, engine, &writer));
+      json += pretty_ ? PrettyPrint(sb.GetString()) : sb.GetString();
+      json += "\n";
+    }
+    return json;
   }
 
-  return json;
+ private:
+  bool pretty_;
+};
+
+constexpr auto JSONString = JSONGenerator{false};
+constexpr auto JSONStringPretty = JSONGenerator{true};
+
+// Both field sets are the worst-case ordering of each other - i.e. the parser cannot
+// reliably predict the next field in B given the definition of A.
+FieldVector TestFields1() {
+  return {
+      field("int", int32()),
+      field("str", utf8()),
+  };
+}
+FieldVector TestFields2() {
+  return {
+      field("str", utf8()),
+      field("int", int32()),
+  };
+}
```

Comment:
I held back on adding fields for now since I wasn't sure how it would affect regression tests (for the ordered case, at least).

I also tried out randomly shuffling the fields for every row. Given the parser's current prediction strategy and limited number of fields, this felt unnecessary - but it would be a more robust solution. The new `JSONGenerator` helper could be be altered to accept a callable rather than constant input.

Reply:
We should indeed try not to modify the characteristics of existing benchmarks. We can however add a different set of benchmarks with a lot more fields, potentially varying fields per line, etc.

Reply:
Alright, I'll add distinct benchmarks for longer schemas - in which case, I'll implement field randomization too.

On another note, it seems worthwhile to get a control case for rapidjson on its own - possibly with a dummy handler. Thoughts?

Reply:
Hmm, why not, that would give an idea of our own overhead.